### PR TITLE
Use asyncio.Event to stop stream in heartbeat route

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

This may fix some of the slowdowns we've seen in spaces loading recently. 

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

**Comment by abidlabs** [comment]
Can you describe what this PR is doing conceptually @freddyaboulton? Not sure I understand it

**Comment by freddyaboulton** [comment]
We need to stop the heartbeat messages when the server is shutting down because starlette waits for all connections to close before exiting.

The old way we were doing that was by polling the is_running prop but that was taking up too much of the event loop. Some gradio spaces like gradio/calculator were taking a really long time to load (30 seconds). 

So we're switching to this event-based approach that works better than polling.

File: gradio/routes.py [hunk 1]
```diff
@@ -606,9 +607,16 @@ async def wait():
                 return "wait"
 
             async def stop_stream():
-                while app.get_blocks().is_running:
-                    await asyncio.sleep(0.25)
-                return "stop"
+                # For some reason tests are failing when using events only on CI
+                # so we use a different approach for CI but
+                # locally and in deployment we should use the event-based approach
```
[the remark this answers is not in the document]
Comment (aliabid94) [reply gradio/routes.py:612]: Agreed. Do you know why it doesn;t work on CI? Ideally we use the event approach everywhere

**Comment by aliabid94** [comment]
Nice PR, the event approach is cleaner but I don't think this should be merged until we know why it doesn't work on E2E tests.

**Comment by freddyaboulton** [comment]
@abidlabs The deployed space is up!

**Comment by freddyaboulton** [comment]
CI is finally green @abidlabs @aliabid94 ! It was a python 3.8 issue

**Review by aliabid94 — APPROVED** [review]
LGTM!

**Comment by freddyaboulton** [comment]
Thanks for the reviews everyone!!

**Pull request merged**